docs/spec_draft.md

```diff
@@ -1444,6 +1445,144 @@ specification. Numeric precision is implementation-defined.
+   * `body` $=$ `reduce_with_select` defined as
+
+     ```c++
+     reduce_with_select(tensor<E> arg0, tensor<E> arg1) -> tensor<E> {
+      return select(arg0, arg1) ? arg0 : arg1;
+     }
+     ```
+     where `E = element_type(operand)`. `select` is transitive, else the
+     behavior of the operation is undefined.
+
+ * Let's define `selected_indices` such that `selected_indices[index]
+   = index-of(selected_values[index], operand)`, where `index` belongs to
+   the index space of `selected_values` and `index-of(a, t)` provides the index
+   of value `a` in tensor `t` as selected by `select` function.
+
+ * `result[i0, i1, ..., iR-1] = reduce(source_values)` where:
+   * `source_values = { source[j0, j1,..., jR'-1]:
+     selected_indices[j0, j1, ..., jR-1] = {0, i1, ..., iR'-1}}`.
+
+   * `reduce(source_values)` = `eval(schedule)` for some binary tree `schedule`
+    where:
+     * `eval(node)` = `scatter(eval(node.left), eval(node.right))`.
+     * `eval(leaf)` = `leaf.value`.
+
+   * `schedule` is an implementation-defined full binary tree whose in-order
+     traversal consists of:
+     * `source_values[index]` values, for all `index` in the index space
+      of `source_values`, in the ascending lexicographic order of `index`.
+      * Interspersed with an implementation-defined amount of `init_value`
+      at implementation-defined positions.
+
+### Inputs
+
+| Name                | Type                                       | Constraints      |
+|---------------------|--------------------------------------------|------------------|
+| `operand`           | tensor of any supported type               | (C1-C9)          |
+| `source`            | tensor of any supported type               | (C1-C3)          |
+| `init_value`        | 0-dimensional tensor of any supported type |                  |
+| `window_dimensions` | 1-dimensional tensor constant type `si64`  | (C1), (C3), (C4) |
+| `window_strides`    | 1-dimensional tensor constant type `si64`  | (C3), (C5)       |
+| `padding`           | 2-dimensional tensor constant type `si64`  | (C3), (C6)       |
+| `select`            | `function`                                 | (C7)             |
+| `scatter`           | `function`                                 | (C4), (C8)       |
+
+### Outputs
+
+| Name     | Type                         | Constraints |
+|----------|------------------------------|-------------|
+| `result` | tensor of any supported type | (C9)        |
+
+### Constraints
+
+  * (C1) rank(`operand`) $=$ rank(`source`) $=$ size(`window_dimensions`).
+
+  * (C2) `operand` and `source` have the same element type.
+
+  * (C3) `dim(source, i) = strided_bound[i]` for any i $\in$ axis(`source`)
+    * `strided_bound[i] = (padded_operand[i] == 0 || window_dimensions[i] > padded_operand[i]) ? 0 : floor((padded_operand[i] - window_dimensions[i]) / window_strides[i]) + 1`.
+    * `padded_operand[i] = padding[i][0] + dim(operand, i) + padding[i][1]`.
```

Comment:
Same feedback as in https://github.com/openxla/stablehlo/pull/354/files#r1030970487. This will also allow to subsume the `rank(source)` part of C1.